catalog/resource_connection_test.go

```diff
@@ -0,0 +1,315 @@
+package catalog
+
+import (
+	"net/http"
+	"testing"
+
+	"github.com/databricks/databricks-sdk-go/apierr"
+	"github.com/databricks/databricks-sdk-go/service/catalog"
+	"github.com/databricks/terraform-provider-databricks/qa"
+	"github.com/stretchr/testify/assert"
+)
+
+func TestConnectionsCornerCases(t *testing.T) {
+	qa.ResourceCornerCases(t, ResourceExternalLocation())
+}
+
+func TestConnectionsCreate(t *testing.T) {
+	d, err := qa.ResourceFixture{
+		Fixtures: []qa.HTTPFixture{
+			{
+				Method:   http.MethodPost,
+				Resource: "/api/2.1/unity-catalog/connections",
+				ExpectedRequest: catalog.CreateConnection{
+					Name:           "testConnectionName",
+					ConnectionType: catalog.ConnectionType("testConnectionType"),
+					Comment:        "This is a test comment.",
+					OptionsKvpairs: map[string]string{
+						"host": "test.com",
+					},
+					PropertiesKvpairs: map[string]string{
```

Comment:
This needs to be a list of options:
```
"options":[
  {"key":"host","value":"mysqk.fakedb.com"},
  {"key":"port","value":"1234"},
  {"key":"user","value":"user123"},
  {"key":"password","value":"password123"}
]
```

Reply:
Right now it is not a list:
"options_kvpairs": {
   "host": "test.com"
}

Reply:
Is it really expected that it will be repeated keys? Other APIs are accepting objects as well, in addition to key/value pairs...

Reply:
I believe the api docs are wrong, this is something we need to correct. There's no difference between the connection api code and other apis.

Reply:
cc @Doug-Luce - I am not sure if the open api spec actually works - the spec mentions the options in a KV pair way:
```
"options_kvpairs": {
   "host": "test.com",
   "port": "1234",
   "user": "user123",
   "password": "password123"
}
```
but I always know this works:
```
"options":[
  {"key":"host","value":"mysqk.fakedb.com"},
  {"key":"port","value":"1234"},
  {"key":"user","value":"user123"},
  {"key":"password","value":"password123"}
]
```
It could be possible that we have an implementation deficiency of the connection API that caused us to not be able to take the kv style of input, but I think most likely it's a bug in our API doc.

Reply:
Please fix the OpenAPI spec, as we (and customers!) depend heavily on its accuracy to be able to use your team's API.

BTW: do we really want this to be a repeated list of KV pairs? We should be critical about these interfaces and think from the customer's perspective, aiming to do what is natural and intuitive where possible, rather than force all customers to adapt to an unusual pattern because there is some small technical issue. In the worst case, you should be able to define your proto using the [Any type](https://protobuf.dev/programming-guides/proto3/#any) and deserialize yourself (scalapb docs here: https://scalapb.github.io/api/com/google/protobuf/any/Any.html).

Reply:
I queried the REST API and saw the following response:
```
<     {
<       "connection_id": "<REDACTED>",
<       "connection_type": "POSTGRESQL",
<       "created_at": 1681945142627,
<       "created_by": "<REDACTED>",
<       "credential_type": "USERNAME_PASSWORD",
<       "full_name": "<REDACTED>",
<       "metastore_id": "<REDACTED>",
<       "name": "<REDACTED>",
<       "options": {
<         "host": "<REDACTED>,
<         "port": "5432"
<       },
<       "owner": "<REDACTED>",
<       "provisioning_info": {
<         "state": "ACTIVE"
<       },
<       "read_only": true,
<       "securable_kind": "CONNECTION_POSTGRESQL",
<       "securable_type": "CONNECTION",
<       "updated_at": 1681945142627,
<       "updated_by": "<REDACTED>",
<       "url": "jdbc://<REDACTED>:5432/"
<     }
```
Options is named `options` and is a dictionary of strings. This response had no properties, but that field is modeled identically internally.

I'll file a ticket to correct the OpenAPI specification. In the meantime, it seems like we should model this as a `map[string]string`.

Reply:
I'm updating the OpenAPI spec to fix the naming of options as well as to reflect that properties_kvpairs is not a part of the response. I'll also add a couple of missing fields.

For creating the connection, options are specified as `map[string][string]`.
This is valid:
```
"options_kvpairs": {
   "host": "test.com",
   "port": "1234",
   "user": "user123",
   "password": "password123"
}
```

Reply:
I think it should still be named options, as that is the name of the nested field inside after inlining?

Reply:
Yes, it would be options and not options_kvpairs. Copy-pasted from Andrew's post, but this is the body I used to create the connection.
```
{
    "name": "openapi_connection_spec_test_2",
    "connection_type": "MYSQL",
    "options": {
        "host": "test.com",
        "port": "5432",
        "user": "dougluce",
        "password": "redacted"
    },
    "properties": {
        "some_prop": "test1",
        "some_prop2": "test2"
    }
}
```